### ISSUE TITLE
SoFurry site redesign will probably break plugin

### ISSUE THREAD

**Issue by Eupeptic1** [Simon-Tesla/RaccoonyWebEx#115]
SoFurry is working on a redesign of their site, variously called SoFurry NEXT or SoFurry X.  It will probably break the existing SoFurry plugin when it goes into production.

Their own news post about it, from June 2021, is here: https://www.sofurry.com/groups/news?id=1&newsId=2195 They don't specify a date when it will go into production.

The news post does mention an API, but I can't readily find any more details about it.  It also promises "responsive design", which may have implications for the plugin working correctly on desktop vs. mobile.

A preview of the new site is available at https://sofurrybeta.com/ .  Only the "login" link shows on that page (at the top right), but a little fooling around without logging in shows that https://sofurrybeta.com/search is probably the main "search" page.

**Comment by Eupeptic1** [comment]
### Analysis

A very quick look at the preview site shows hash-type submission IDs - not integers anymore.  The submission files are referenced by a GUID with no filename extension.

There is a "download" link on the site for *some* images; I'm not sure what the criteria are for providing it.  If it's there, I think you have to be logged in for it to work.

There appears to be a "multiple image" or "comic" submission type; see https://sofurrybeta dot com/s/qB1QE1j6 for an example.  It seems to have a single submission ID with next/prev buttons to view the images.  All of the images appear to be linked from a `<div id="comic-view">` towards the end of the source HTML.  Support for this submission type may end up being related to issue #111 .

The stories appear to be in-line HTML.  Like the images, some of them have a "download" link and some don't.

Music files appear to be self-hosted, and not on Soundcloud or similar.  There are links in the page source to the mp3 or whatever (but again, as a GUID with no file extension).

The one video available seems to be self-hosted.  There is a link in the page source to the WebM.

The 3D model I checked seems to be self-hosted.  There is a link in the page source to the file.  The file format appears to be this: https://en.wikipedia.org/wiki/GlTF .  (It starts with `glTF`, then some control characters, then some JSON that mentions that it came out of Blender, then a big wad of what looks to be 4 or 8 byte ints or floats.)

**Comment by Eupeptic1** [comment]
I have a real basic version of a plugin for the new site working.

Building a list of files for "open all in tabs" usually works.  Downloading most art and photos works.  Comic-book art doesn't work - I figure it should at least download the first page/first image, but it doesn't do that yet.  Music, video, and 3D don't work.

Filename extensions are just not available on the site.  I am leaving it unset in the plugin, and letting the existing code in `siteActions.ts` figure it out.  I *think* that usually involves fetching the file again, which might be an issue if the user opens lots of tabs?  Dunno.  The browser *should* already know what the MIME type was when it downloaded the image when the page loaded in the first place, but I couldn't come up with a way to query the browser for that.

The site doesn't seem to be single page app, but sometimes I have to hit "reload" to get the plugin to notice that there's actually something it can parse on a page.  Not all the time, though, so I'm not yet sure why that is.

Open questions:

1.  It looks like the submission filenames are going to be GUIDs, which are less than useful for later browsing of downloaded files.  Should the plugin attempt to generate a more reasonable filename, like the e621 plugin does?  Or, is it possible to ship Raccoony with "use global" unchecked for "Download path" for just this plugin, and set it to something like `{submissionId}_{title}_by_{author}.{extension}`?

2.  When SoFurry puts the new site in production, will the new site be the only one available to users?  Or will users be able to pick, like the old and new UIs on FA?

**Comment by Simon-Tesla** [comment]
1. The siteFilename logic is kind of a hack for e621 to begin with. At this point I feel like using the guid plus an extension is fine. In general, as long as siteFilename is likely to be unique, I don't care about exact formats. The existing hacks can stay, but I don't want to create new ones if I can help it.

Typically responsive design means that they're using the same site for both desktop and mobile, and the CSS/JS is written to accommodate varying screen sizes with a design that is appropriate to that size. The new FA UI is also responsive in this manner, as I recall, so unless they do something really odd, more likely than not the code written for desktop should just work on mobile. Assuming any mobile browser ever supports Raccoony again; until then, the issue's kinda moot.

The issue with not having extensions is concerning, though not exactly unique to SoFurry as I recall. It's an area I need to revisit; I'd hoped my fetch logic would use the cached version and not try to refetch from the server, but I'd need to verify whether that is the case or not as I suspect it may not be.

There may also be some alternative methods for figuring out a content type that don't involve a refetch, it's an area we'd need to look into again.

**Comment by Eupeptic1** [comment]
> The existing hacks can stay, but I don't want to create new ones if I can help it.

OK.

> Assuming any mobile browser ever supports Raccoony again; until then, the issue's kinda moot.

Firefox 68 in an Android VM, with an unsullied copy of the VM disk image in reserve.  When Firefox gets pwnt, push the "nuke and pave" button to shut down the VM, delete the working copy of the disk image, make a copy of the pristine image, and start the VM again.  :D

> I'd hoped my fetch logic would use the cached version and not try to refetch from the server, but I'd need to verify whether that is the case or not as I suspect it may not be.

I don't know for sure that it refetches from the server; I didn't look at the wire while browsing.  I will try that and report my findings.

> There may also be some alternative methods for figuring out a content type that don't involve a refetch

What I've found so far pretty much says to do what you are already doing - ask the browser to fetch it again, which then allows you to get the MIME type from the headers.

I found a couple of references to *Javascript* having a `.mimetype` property on some elements, including images, but I can't figure out if a) it used to exist but doesn't anymore or b) it exists but you can't get at it from Typescript.

(I also found a bunch of stuff about how to figure out the MIME type of a file on *local* disk, for when you're trying to *upload* a file, but that's not the situation we have.)

If you load a page in Firefox, go to Tools > Page Info, and select "Media", it lists a "Type:" for every image.  I'm just not sure yet if there's a way for an extension to ask for that info.

### New question:

The individual image pages don't seem to use a thumbnail or preview image; they just link to the original-sized image and let it get rescaled in HTML/CSS to fit the window.  The original-sized image has a URL resembling `cdn.sofurry.com/full/8675309-420-69-CE1704`.

The "browse" and "search" pages *do* have thumbnails for their results, and the URLs seem to use the same GUID as the full-size image, but with a different (and consistent) path, like `cdn.sofurry.com/thumbs/8675309-420-69-CE1704`.

When we're on an image page, is it OK to **assume** that the thumbnail exists on the server, and do something like `s/full/thumb/` on the image URL to *generate* a thumbnail URL?  Or should we avoid generating a URL that we don't know for sure exists - in other words, a URL that we haven't seen in the HTML of the image page itself?

**Comment by Simon-Tesla** [comment]
> When we're on an image page, is it OK to assume that the thumbnail exists on the server, and do something like s/full/thumb/ on the image URL to generate a thumbnail URL? Or should we avoid generating a URL that we don't know for sure exists - in other words, a URL that we haven't seen in the HTML of the image page itself?

My preference is against url generation in general unless there's no other way to accomplish something. In any case 'previewUrl' isn't used for thumbnails, it's used for the smaller version of images that many sites provide instead of the original image that's smaller but not thumbnail-sized. The correct thing to do here would be to use the regular URL for the picture that we already see. You'll notice this implementation in the e621 plugin, for example.

**Comment by Eupeptic1** [comment]
> The correct thing to do here would be to use the regular URL for the picture that we already see. 

OK.  I'll do it that way.

**Comment by Eupeptic1** [comment]
**tl;dr**: First cut on the [sofurry-beta-support branch](https://github.com/Eupeptic1/RaccoonyWebEx/tree/sofurry-beta-support) in my repo.  Has basic functionality on desktop Firefox for everything but stories; I haven't tested any other platforms.

#### Long version:

I've been working on it some more, and it works well enough to put on Github.

#### Download support:

Normal images (both art and photos) - working.
Comics - can download cover or first image only.  The site allows multi-image submissions, which probably needs support in the rest of Raccoony.
Music - can download first track only.  There is one example of a multi-track submission, which probably needs support in the rest of Raccoony.
Video - working.  URL is generated; see below.
3D models - working.  File extension is forced; see below.
Stories - no support; see below.

The music and video players seem to be loaded by other scripts on the page, so they're not always there by the time Raccoony looks at them, so looking for the file linked from the player didn't always work.  For music, there's a static element in the page that also links to the music file, so I use that.  For video, there's a static thumbnail in the page that has the same GUID as the video file, so I generate the video file URL from that.

The 3D models come through as GlTF files.  The existing extension-finding code in `siteActions.ts` doesn't know what to do with them, so I force an extension of `.glb`, which I *think* is correct.  I thought about modifying the extension-finding code, but SoFurry doesn't provide the right MIME type when you try to download a 3D model (it's just application/octet-stream), so forcing the extension seemed easier.  One consequence of this is that the JSON in the metadata file says `"type": "unknown"`.

I detect stories but don't do much else with them.  I had the brain wave of trying to build a `data:` URL out of the story HTML, and feeding that to the rest of Raccoony, but it didn't seem to work.  :)  The site doesn't appear to support PDF, Word, RTF, etc - just stories as HTML.

#### Open-in-tabs support:

Works from the 'all' page, but not from the other pages.  Needs more work.

#### Single-page-appy stuff:

If I'm looking at a search page, and I click on one of the submissions, the page changes to show the submission, but Raccoony doesn't run; if I then hit "reload", Raccoony runs.  If I'm looking at a search page, and I right-click a submission and open it in a new tab, Raccoony runs.  Not sure why.

#### From previous messages:

The browser cache seems to do the right thing; if `ensureMediaHasFilenameAndExtension()` has to do a fetch to get the MIME type, it doesn't hit the server again if the item in question is still in the browser cache.  Tested on three images by starting Wireshark, loading an image, clicking the Raccoony "download" button, stopping Wireshark, and comparing the traffic size to the image size.  The traffic size was less than double the image size, so I concluded that the browser wasn't downloading the whole image from the server again.

**Comment by Eupeptic1** [comment]
**tl;dr**: I reworked the plugin to use a mutation selector, like Deviant Art and Pixiv, but it doesn't work correctly all the time, and I'm not sure why.  Sometimes the mutation selector fires and sometimes it doesn't, even when the page has clearly changed.
Update on the [sofurry-beta-support](https://github.com/Eupeptic1/RaccoonyWebEx/tree/sofurry-beta-support) branch in my repo.

#Long version

Open-all-in-tabs works better now; the problem there is that the markup has slightly different classes for artwork, photos, stories, etc, and my querySelector was too specific.  Making the query more general makes Raccoony offer open-in-tabs on all types of search-result pages.

The next problem I tried to fix is that you always get the raccoon head if you come in from the beginning to one of the search pages, such as https://sofurrybeta.com/search , https://sofurrybeta.com/search/Artwork , https://sofurrybeta.com/search/Writing , and so on.  However, if you're looking at one of those pages, and click the link near the top of the page to switch to a different type of search, the raccoon head disappears.  (For example, if you're on https://sofurrybeta.com/search/Artwork, and then click on the "Writing" link right under the search box, the page of search results appears, but no raccoon head.)

I thought Sofurry might be rewriting parts of the page, instead of triggering a complete new page load, so I added a mutation selector, like Deviant Art and Pixiv have.  I tried to make it trigger on `div.row`, which is the search results grid, but that didn't work; triggering on `body` at least makes it work sometimes.  Once it stops working, I can't get it working again without a hard reload (shift-reload).

This is what I see right now:

1.  In a new tab, go to http://sofurrybeta.com/search .
2.  Raccoony runs and makes a list of all the search results.  The raccoon-head appears and offers the "open all in tab" button.
3.  Click the "artwork" link under the search bar.
4. Raccoony sees a mutation of type `childList`.  It runs again and makes a new list of the new results.  (You can tell it's seeing a different set by either inspecting the URLs, or just by the count of links it finds.)  It gets as far as saying `[raccoony] hasPageLinks true` in the console, but the raccoon head never appears.
5. Click on another link under the search bar, like "writing" or "music".
6. Raccoony does not see the mutation and does not seem to run further; the raccoon head does not appear again.

It seems like once you lose the raccoon head, you can't get it back, except by doing a shift-reload on whatever page you are on.

It *sort of* feels like a race condition or other timing problem when Raccoony gets called by the mutation selector.  Like, the plugin is running, and feeding the right data to main Raccoony, but main Raccoony is injecting the raccoon head on the old page, or possibly on the new page but then it gets overwritten by SoFurry, so the raccoon head doesn't end up displaying.  I am not sure how to fix this situation.

**Comment by Simon-Tesla** [comment]
Hm... I'll have to take a look, but if it's a Single Page App then I suspect that it'll require changes a bit more fundamental than the can be handled at the plugin level. FN is the only other SPA-style site that I've attempted to support and it's also somewhat buggy IIRC. 

Oddly enough, it looks like they're fetching a new page HTML when navigating (e.g. going from the search results to submission) and then replacing *the entire page HTML* with it using their script which is an, uh, *interesting* strategy. It'd explain why the change hook on body doesn't really work, since presumably that gets destroyed along with the rerender, which appears to eat the entire DOM. I wonder if they plan on keeping it this way for the released version, or if this is something they're doing to make testing changes faster.

That said, given how long I've seen other furry gallery sites stay in beta, I'm not too concerned about the urgency on this, especially since they haven't even technically opened to the general public, and I'm not overly inclined to chase a potentially moving target. Maybe once they hit public beta phase.

**Comment by Eupeptic1** [comment]
> FN is the only other SPA-style site that I've attempted to support and it's also somewhat buggy IIRC.

I haven't tried to use Raccoony on FN, but I know it does something similar; if you run `wget` or similar on a single FN submission, you get some headers with a little bit of metadata, and then a blank spot that would be filled in if all the Javascript could run.

I tried the mutation selector for SoFurry just because that made the experimental Pixiv support jump from "pretty broken" to "kind of works sometimes", and I *thought* SoFurry was doing something similar to Pixiv.

> It'd explain why the change hook on body doesn't really work

Good, I'm not crazy.  :D

> I wonder if they plan on keeping it this way for the released version, or if this is something they're doing to make testing changes faster.

I don't know.  The beta site appears to have maybe 100-200 submissions copied from the production site, and it probably doesn't get that many views... they may change the structure to something simpler once they open it up to more users and see what the traffic levels are like.

> Maybe once they hit public beta phase.

I went back and re-read their announcement post from June 2021, and the comments on it (linked in the original issue here).  Apparently it's been in beta for at least five years, so I agree that it's probably not going to be deployed imminently.  It looks like the dev team is probably quite small.

I think for now I'll leave it on the branch in my repo, and maybe occasionally check SoFurry's news page to see if it's getting any closer to production.

**Comment by Simon-Tesla** [comment]
Sounds good! Thanks for looking into this though. :)

**Comment by Eupeptic1** [comment]
SoFurry posted another [status update](https://www.sofurry.com/groups/news?id=1&newsId=2220) in December 2021.  Still no ETA for when the redesigned site might go into production.

The status post and the comments show that there are some features (like a "popular" page or category) that are not yet implemented on the beta site, so further changes to the Raccoony plugin may be required as those become available.

The beta site now claims to have 673 submissions of all kinds.  It looks like a few artwork submissions have been added since I was testing in September/October 2021, but nothing in the other categories.

**Comment by Eupeptic1** [comment]
SoFurry posted another [status update](https://www.sofurry.com/groups/news?id=1&newsId=2241) in May 2022.  Still no ETA for when the redesigned site might go into production.

I tried my existing plugin on the beta site again and it still works like it did back in October 2021 - you get the raccoon-head menu on the first gallery of images, or on the first image you click on.  After that, the raccoon head goes away, and the only way to get him back is to reload the page in your browser.

The dev does seem to be answering user-type questions on that status update.  Do you think it would be worthwhile to ask about things like them replacing most/all of the page HTML at some point during navigation?

**Comment by Simon-Tesla** [comment]
Closing this for now. Let's reopen this when we have a solid ETA/beta phase.

**Comment by Eupeptic1** [comment]
The place to watch for updates seems to be https://www.sofurry.com/groups/news?id=1 ; look for the "SoFurry NEXT Dev Update" threads.

When a new news item is posted there, I think it also appears on the home page https://www.sofurry.com/ for a while.
